nvflare/app_common/ccwf/client_controller_executor.py

```diff
@@ -0,0 +1,382 @@
+
+                self.controller.stop_controller(fl_ctx)
+
+                self.log_info(fl_ctx, f"Finished control flow {self.controller.name}")
+
+                self.update_status(action=f"finished_{controller_id}", error=None, all_done=True)
+
+            self.update_status(action=f"finished_start_task", error=None, all_done=True)
+
+            return make_reply(ReturnCode.OK)
+
+        elif task_name == self.report_final_result_task_name:
+            return self._process_final_result(shareable, fl_ctx)
+
+        else:
+            self.log_error(fl_ctx, f"Could not handle task: {task_name}")
+            return make_reply(ReturnCode.TASK_UNKNOWN)
+
+    def _get_status_report(self):
+        with self.status_lock:
+            status = self.current_status
+            must_report = False
+            if status.error:
+                must_report = True
+            elif status.timestamp:
+                must_report = True
+
+            if not must_report:
+                return None
+
+            # do status report
+            report = copy.copy(status)
+            return report
+
+    def update_status(self, last_round=None, action=None, error=None, all_done=False):
+        with self.status_lock:
+            status = self.current_status
+            status.timestamp = time.time()
+            if all_done:
+                # once marked all_done, always all_done!
+                status.all_done = True
+            if error:
+                status.error = error
+            if action:
+                status.action = action
+            if status.last_round is None:
+                status.last_round = last_round
+            elif last_round is not None and last_round > status.last_round:
+                status.last_round = last_round
+
+            status_dict = status.to_dict()
+            self.logger.info(f"updated my last status: {status_dict}")
+
+    def _process_final_result(self, request: Shareable, fl_ctx: FLContext) -> Shareable:
+        peer_ctx = fl_ctx.get_peer_context()
+        assert isinstance(peer_ctx, FLContext)
+        client_name = peer_ctx.get_identity_name()
+        result = request.get(Constant.RESULT)
+
+        if not result:
+            self.log_error(fl_ctx, f"Bad request from client {client_name}: no result")
+            return make_reply(ReturnCode.BAD_REQUEST_DATA)
+
+        if not isinstance(result, Learnable):
+            self.log_error(fl_ctx, f"Bad result from client {client_name}: expect Learnable but got {type(result)}")
+            return make_reply(ReturnCode.BAD_REQUEST_DATA)
+
+        self.log_info(fl_ctx, f"Got final result from client {client_name}")
+
+        fl_ctx.set_prop(AppConstants.GLOBAL_MODEL, result, private=True, sticky=True)
+
+        if self.persistor:
+            self.persistor.save(result, fl_ctx)
+        else:
+            self.log_error(fl_ctx, "persistor not configured, model will not be saved")
```

Comment:
Maybe we can add some additional enhancement saying if users really want to persist things. (like persist_per_round or something if set to 0 means no persist)
If they did not specify that then no persistor is not an error?

Reply:
I can change to a warning, as this persistor is only used when distributing the final results

Reply:
maybe change this to warning is better